### ISSUE TITLE
Sampling gets stuck after repeated instances of "UserWarning: Random walk proposals appear to be extremely inefficient."

### ISSUE THREAD

**Issue by asb5468** [joshspeagle/dynesty#160]
Hello, I'm using dynesty for an 8-dimensional hierarchical inference problem. I have launched a set of 100 runs with the same likelihood and priors (different data), but for most of the runs, they seem to run indefinitely after throwing "UserWarning: Random walk proposals appear to be extremely inefficient. Adjusting the scale-factor accordingly." I tried restarting with reflective boundary conditions, but those runs have also gotten stuck. None of the remaining ~60 runs have been able to accept new points for over a week (the runs that did finish did so in about a day), and they got stuck in a range of dlogZs between ~0.15-7. I tried digging into where this happens in the code, and it looks like the scale factor is reduced each time this warning comes up. To me it seems like this would be isolating the sampler into a false maximum, since if it were narrowing in on the true peak, a new point would have been accepted or the dZ threshold would have been passed. Is there anything I can do to bypass this problem? My likelihood isn't abnormally peaked, and the posteriors in the trace plots look reasonable but wider that I'd expect for a finished run. Thank you!

**Comment by joshspeagle** [comment]
>  I tried digging into where this happens in the code, and it looks like the scale factor is reduced each time this warning comes up. To me it seems like this would be isolating the sampler into a false maximum, since if it were narrowing in on the true peak, a new point would have been accepted or the dZ threshold would have been passed. Is there anything I can do to bypass this problem?

You are 100% correct that repeated instances of this behavior would cause the sampler to get stuck in a false maximum. The original intent of this was that this generally would happen either close to an edge or in cases where the covariance structure has been misestimated, so shrinking the proposal allows a sample to be proposed from the current live point so that sampling will hopefully proceed as normal with the other points. In practice, however, you sometimes get behavior such as the one you are finding.

In terms of getting around this, in theory using another sampling method might resolve this (e.g., `'unif'`), but that's not really much of a fix. This issue could actually evidence for a bug elsewhere: there appear to be some issues that have been hard to squash when estimating covariances, so this might just be "my covariance always makes me propose `nan`s so I can't accept anything and get stuck in an endless loop". Knowing whether that's happening or not would be useful here if you can pull out some of that information for those runs. The other fix would also be on my end: I could try and rewrite some of the internals of the code to allow for "restarts" so that after failing to propose a new position from a random live point you just pick a new one at random and try again. I *think* such a change to the internals would be possible.

**Comment by asb5468** [comment]
Thanks for the quick reply! I did take a look at the values of the likelihood and parameters that it was trying, and the log-likelihoods were finite in a narrow range with a scatter of around 4. The parameters weren't very near the true values that I used when simulating the data. I think this means it's not getting stuck in a wonky part of the parameter space that produces `nan`s. In the meantime I'll try with uniform sampling!

**Comment by joshspeagle** [comment]
Okay, good to know. If this problem continues to arise I am happy to troubleshoot fixes. Best of luck!

**Comment by nikhil-sarin** [comment]
Hiya I am also experiencing the same issue, with a similar diagnosis as @asb5468, I'll try the 'unif' method but an internal fix to randomly choose a different point would be great!

**Comment by joshspeagle** [comment]
Okay, I should definitely look into this sooner rather than later then. It seems more people have been having this issue after the overhaul associated with PRs refactoring the periodic and reflective boundary conditions, so I should check everything is working as intended. I’ll try to get to this ASAP.

**Comment by joshspeagle** [comment]
Okay, I don't see any obvious bugs in the [`rwalk` method](https://github.com/joshspeagle/dynesty/blob/master/dynesty/sampling.py#L97), but if either of you have a copy of the `results` object saved from one of these stalled runs that you can send me that would be extremely helpful diagnosing any possible problems that I might have missed in the actual proposal (new eyes tend to catch things) or possible problems that might be occurring upstream.

Regardless, introducing a new feature to allow for retries (with an associated warning) should be a good addition. Functionally, I should be able to add in a quick check [here](https://github.com/joshspeagle/dynesty/blob/caf952602eb44d9e0a0d01f535b83d3a40f5a6da/dynesty/sampler.py#L304) when new points are proposed that checks whether an evaluation "failed" (based on a new flag I can add if, say, the method tries and fails to find a new point after `N` attempts, where `N` is say 5) and removes it from the queue if so, then restarts the whole process of proposing new points if the queue is empty. This should work well with parallelized proposals by just removing any particular ones that failed without requiring a full replacement of the associated queue. Sound reasonable?

**Comment by asb5468** [comment]
Yep, I was thinking of a similar work-around, but I guess that `N` should be chosen so that this wouldn't cause the logic to backfire in the original cases you envisioned for reducing the scale factor of the ellipse. I've attached a compressed pickled resume file (can be unpickled with python2.7) for one of my runs that got stuck like this. Let me know if you have trouble opening it!
[power_law_99_hyper_dyn_resume.pickle.zip](https://github.com/joshspeagle/dynesty/files/3762928/power_law_99_hyper_dyn_resume.pickle.zip)

**Comment by joshspeagle** [comment]
Okay, I've played around with the results for a bit. I think part of the problem here is that some of the parameters appear to remain totally unconstrained, and those that are constrained appear to be hitting the edge. I've just plotted two of them below.
![image](https://user-images.githubusercontent.com/13280133/67521285-95b9c900-f678-11e9-925b-401c410a2183.png)
![image](https://user-images.githubusercontent.com/13280133/67521525-12e53e00-f679-11e9-9b65-ad4816b9e180.png)

This looks like it's messing with the ellipsoid decomposition and proposals, as seen in the distribution of live points.
![image](https://user-images.githubusercontent.com/13280133/67521246-7fac0880-f678-11e9-8c3e-5cb0faea74c6.png)

The scale factor then attempts to adjust to compensate.
![image](https://user-images.githubusercontent.com/13280133/67521435-e2050900-f678-11e9-96ea-09355aea87c3.png)

What really appears to be the issue is that either a few of the decompositions are *terrible* or the associated live points are on the top of an island of absolutely amazing log-likelihood (most likely the former). So while things run OK most of the time, you indeed get these humongous spikes where it stalls:
![image](https://user-images.githubusercontent.com/13280133/67521678-58097000-f679-11e9-9e82-88bd7022fa76.png)

So I definitely can confirm the problem. I think this is a result of the ellipsoid decomposition not performing well on this problem (and/or being unstable, or both), leading to occasionally awful proposals, and the code attempting to compensate. Introducing retries should definitely resolve this, so I'll try and patch that in at some point soon.

Thanks a lot for sending those results my way!

**Comment by joshspeagle** [comment]
I'm closing this for now. Hopefully the changes in #163 and resolving #78 and #140 should also fix this.

**Comment by 3fon3fonov** [comment]
I am afraid I have to re-open this. With dynesty 1.2, (Dynamic nested sampling, rwalk, multi bound, posterior frac = 1.0, dlogz stop=0.1) I am way too often seeing the following (example):

```
22804it [32:19,  1.94it/s, batch: 0 | bound: 278 | nc: 25 | ncall: 596613 | eff(%):  3.818 | loglstar:   -inf < -13685.908 <    inf | logz: -1322871it [32:19,  3.07it/s, batch: 0 | bound: 280 | nc: 25 | ncall: 599533 | eff(%):  3.811 | loglstar:   -inf < -13675.915 <    inf | logz: -1322895it [32:30,  3.07it/s, batch: 0 | bound: 280 | nc: 148 | ncall: 600342 | eff(%):  3.810 | loglstar:   -inf < -13670.159 <    inf | logz: -13713.353 +/-  8.960 | dlogz: 401.066 >  0.100]../../exostriker/lib/dynesty_1_2/sampling.py:243: UserWarning: Random walk proposals appear to be extremely inefficient. Adjusting the scale-factor accordingly.
  warnings.warn("Random walk proposals appear to be "
../../exostriker/lib/dynesty_1_2/sampling.py:243: UserWarning: Random walk proposals appear to be extremely inefficient. Adjusting the scale-factor accordingly.
  warnings.warn("Random walk proposals appear to be "
../../exostriker/lib/dynesty_1_2/sampling.py:221: UserWarning: Random number generation appears to be extremely inefficient. Adjusting the scale-factor accordingly.
  warnings.warn("Random number generation appears to be "
22896it [43:44,  4.77s/it, batch: 0 | bound: 280 | nc: 25 | ncall: 600367 | eff(%):  3.810 | loglstar:   -inf < -13670.008 <    inf | logz: -1322947it [43:44,  3.23s/it, batch: 0 | bound: 282 | nc: 25 | ncall: 608048 | eff(%):  3.770 | loglstar:   -inf < -13660.710 <    inf | logz: -1322990it [44:01,  3.23s/it, batch: 0 | bound: 283 | nc: 25 | ncall: 611052 | eff(%):  3.759 | loglstar:   -inf < -13655.916 <    inf | logz: -13698.388 +/-  8.960 | dlogz: 445.853 >  0.100]
```

As you can see it works, then it goes to the error, but after some time it starts sampling again. However, often it stucks to infinite number of the UserWarning, and the sampler stacks. This is mostly happening at the final stages when in constructs the posteriors. The only solution I found is to observe the ncall before it sucks, and reset with maxcall=ncall before it stucks, but this is super time consuming. Some of my setups are running for days on a 40 cpu machine, so you can imagine the forstration when I have to restart the run.

Also, for my problem(s) I find that rwalk is the fastest and most efficient sampling option, thus I would really appreciate help with this.

**Comment by segasai** [comment]
@3fon3fonov  You are clearly not using the latest released version as  1.2.3 version does not have that warning

**Comment by 3fon3fonov** [comment]
@segasai, I must admit, this is true! I realized this after I post above.... So it seems updating to 1.2.3 is working fine now, but I am still experimenting. Let's see how this will go :) Else, sorry for the spam!
